### PR TITLE
minimal fixes to run DataCollatorForWholeWordMask with return_tensors="np" and return_tensors="tf"

### LINKED ISSUE #13890: DataCollatorForWholeWordMask does not work with return_tensors = "np" and return_tensors = "tf"
## Environment info
- `transformers` version: 4.11.2
- Platform: Linux-4.19.0-14-cloud-amd64-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.5
- PyTorch version (GPU?): 1.9.1+cu102 (False)
- Tensorflow version (GPU?): 2.6.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help
@Rocketknight1 

## Information
The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below)

I'd like to use `DataCollatorForWholeWordMask` for masked language modeling in TensorFlow.

##### `return_tensors = "np"`

```
from transformers.data import DataCollatorForWholeWordMask
from transformers import BertTokenizerFast

return_tensors = "np"
tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")
collate_fn = DataCollatorForWholeWordMask(tokenizer, return_tensors=return_tensors, mlm=True)
collate_fn([{"input_ids": list(range(10))}, {"input_ids": list(range(10))}])
```
```TypeError: numpy_mask_tokens() got an unexpected keyword argument 'special_tokens_mask'```

This is at least partly due to not implementing `DataCollatorMixin.numpy_call` in `DataCollatorForWholeWordMask` (`np_call` is implemented instead). There are also some issues with using TensorFlow tensors in the `np_call` function.

##### `return_tensors = "tf"`

```
from transformers.data import DataCollatorForWholeWordMask
from transformers import BertTokenizerFast

return_tensors = "tf"
tokenizer = BertTokenizerFast.from_pretrained("bert-base-uncased")
collate_fn = DataCollatorForWholeWordMask(tokenizer, return_tensors=return_tensors, mlm=True)
collate_fn([{"input_ids": list(range(10))}, {"input_ids": list(range(10))}])
```

```AttributeError: 'tensorflow.python.framework.ops.EagerTensor' object has no attribute 'clone'```

`inputs.clone()` appears to have been copied from `torch_call` which is not valid in TensorFlow.

### PR DESCRIPTION
**Pull request by dwyatte**
# What does this PR do?

This PR addresses https://github.com/huggingface/transformers/issues/13890 with the minimal fixes to run `DataCollatorForWholeWordMask` with `return_tensors="np"` and `return_tensors="tf"`

Specific problems addressed:
* Renamed `np_call` -> `numpy_call`
* Call `_numpy_collate_batch` instead of` _tf_collate_batch` when returning numpy tensors
* Fix size of `random_words` in `numpy_mask_tokens`
* Clone tensorflow tensors with `tf.identity(tensor)`
* Use tensorflow tensors’ built-in iteration instead of attempting to convert to list
* Change calls to `tf.convert_to_tensor` with `dtype=tf.bool` to `tf.cast` with `dtype=tf.bool`

I've only added a simple test to check for regressions vs all of the padded/unpadded cases as is done for `DataCollatorForLanguageModeling`

Fixes https://github.com/huggingface/transformers/issues/13890

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

CC @Rocketknight1 (looks like you did the initial Numpy/TF implementation of these data collators)

### CODE REVIEW

**Comment by dwyatte** [comment]
Bump @Rocketknight1 @LysandreJik @sgugger since `transformers` releases frequently and it would be nice to have this functionality.

This gives a vetted source for masked language modeling that is not https://github.com/google-research/bert/blob/master/create_pretraining_data.py or https://github.com/tensorflow/models/blob/master/official/nlp/data/create_pretraining_data.py in a package many people are going to be using for modeling (e.g., I could delete my code that is a copy/paste of the former and use integrate against this instead for creating a large offline dataset).

**Review by sgugger — APPROVED** [review]
This looks good to me, thanks for the fixes!

File: src/transformers/data/data_collator.py [hunk 1]
```diff
@@ -996,15 +996,15 @@ def tf_mask_tokens(self, inputs: Any, mask_labels: Any) -> Tuple[Any, Any]:
...
         # We sample a few tokens in each sequence for masked-LM training (with probability args.mlm_probability defaults to 0.15 in Bert/RoBERTa)
 
         masked_indices = tf.cast(mask_labels, tf.bool)
 
         special_tokens_mask = [
-            self.tokenizer.get_special_tokens_mask(val, already_has_special_tokens=True) for val in labels.tolist()
+            self.tokenizer.get_special_tokens_mask(val, already_has_special_tokens=True) for val in labels
         ]
-        masked_indices = masked_indices & ~tf.convert_to_tensor(special_tokens_mask, dtype=tf.bool)
+        masked_indices = masked_indices & ~tf.cast(special_tokens_mask, dtype=tf.bool)
```
Comment (Rocketknight1) [inline src/transformers/data/data_collator.py:1007]: Does this change assume the tokenizer is set to `return_tensors="tf"`?

**Comment by Rocketknight1** [comment]
Sorry for the delay, and thank you for the ping! This looks good to me. I've got one question about the `convert_to_tensor` versus `cast` call, but other than that this is a solid and necessary PR.

**Review by Rocketknight1 — APPROVED** [review]

File: src/transformers/data/data_collator.py (hunk 1 above)
Comment (dwyatte) [reply src/transformers/data/data_collator.py:1007]: Nope, TensorFlow coerces the python/numpy value into a Tensor here:

```
>>> tf.cast([0, 1, 0, 1], dtype=tf.bool)
<tf.Tensor: shape=(4,), dtype=bool, numpy=array([False,  True, False,  True])>
>>> tf.cast(np.array([0, 1, 0, 1]), dtype=tf.bool)
<tf.Tensor: shape=(4,), dtype=bool, numpy=array([False,  True, False,  True])>
```

This behavior is also tested with whatever the default `return_tensors` (looks like they are just lists)
https://github.com/huggingface/transformers/pull/13891/files/94a3cf08f1e328202739233b1d58980edcddc2af#diff-3d06edcb84efb377e8c1b5545f396cdf69a1de21ff82cbf2aab3902e4b42dec5R505-R506

Comment (Rocketknight1) [reply src/transformers/data/data_collator.py:1007]: Ah, that works! The default return_tensors does indeed just return Python lists.

**Pull request merged**